### PR TITLE
Toggle visibilty of Dashboard Editor

### PR DESCRIPTION
**Pull request by SourceDoctor**
Toggle Visibility of Dashboard Editor. So more space for Dashboard is available.
It's a user specific setting.

image

image

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by Jellyfrog** [comment]
This will be confusing to find imo

**Comment by SourceDoctor** [comment]
i'm open for a better place, but thinking to have this "row" also for dashboard would be useful.
Per default Dashboard Editor is visible.

**Comment by SirMaple** [comment]
Just an idea. The option could be put in one of the menus at the top.

Perhaps "Overview -> Dashboard -> Show/Hide Dashboard Editor"

Or even make the editor it's own page then the menu could be "Overview -> Dashboard -> Dashboard Editor"

**Comment by SourceDoctor** [comment]
yeah, could be better there:
"Overview -> Dashboard -> Show/Hide Dashboard Editor"
good point

**Comment by SourceDoctor** [comment]
changed position:

image

image

**Comment by SourceDoctor** [comment]
@Jellyfrog 
now better? :-)

**Comment by murrant** [comment]
Any thoughts to a menu that contains page specific items?

**Comment by SourceDoctor** [comment]
i have some thoughts about showing user created/accessiable Dashboards in this submenu also.
so show/hide Editor is the last Menupoint.

Do you mean this?

**Comment by SourceDoctor** [comment]
now all accessiable Dashboards as submenu:

image

**Comment by murrant** [comment]
That looks better, but one consideration.  When you make clickable menu entry a sub-menu it has issues on mobile.

**Comment by SourceDoctor** [comment]
good point,
as long we've got no mobile GUI, i remove clickability from menu entry

**Comment by SourceDoctor** [comment]
@Jellyfrog @murrant 
something else? :-)

File: app/Http/Controllers/DeviceGroupController.php [hunk 1]
```diff
@@ -33,6 +35,8 @@ public function index()
         return view('device-group.index', [
             'device_groups' => DeviceGroup::orderBy('name')->withCount('devices')->get(),
             'ungrouped_devices' => $ungrouped_devices,
+            # FIXME: this shouldn't be needed here
```
Comment (Jellyfrog) [inline app/Http/Controllers/DeviceGroupController.php:38]: Will you fix these before merging?

Comment (SourceDoctor) [reply app/Http/Controllers/DeviceGroupController.php:38]: oh i oversees this,
with the "new" Gui i have to check this again, if it's still needed here.
Hopefully no :-)

Comment (SourceDoctor) [reply app/Http/Controllers/DeviceGroupController.php:38]: that's a bigger problem i think
this happens now on 

- global Settings
- Poller
- About
- (maybe more)

looks like this happens on laravel designed pages only.

i'll take a look over it,
if you have an idea where to define variable 'hide_dashboard_editor' global for all
Laravel based pages i will be happy to read it :-)

Comment (SourceDoctor) [reply app/Http/Controllers/DeviceGroupController.php]: found a fix

**Comment by SourceDoctor** [comment]
@Jellyfrog found a way to solve the problem

**Comment by SourceDoctor** [comment]
@Jellyfrog 
on my view it's ready for merging ...

File: app/Http/Controllers/LegacyController.php [hunk 2]
```diff
@@ -51,6 +52,8 @@ public function index(Request $request, Session $session)
         $vars['page'] = basename($vars['page'] ?? '');
         if ($vars['page'] && is_file("includes/html/pages/" . $vars['page'] . ".inc.php")) {
             require "includes/html/pages/" . $vars['page'] . ".inc.php";
+        } elseif (Config::has('front_page') && is_file('includes/html/' . Config::get('front_page'))) {
```
Comment (Jellyfrog) [inline app/Http/Controllers/LegacyController.php:55]: Bad rebase

File: resources/views/layouts/menu.blade.php [hunk 3]
```diff
@@ -25,8 +25,26 @@
...
+                                <li><a href="overview?dashboard={{ $dashboard->dashboard_id }}"><i class="fa fa-tv fa-fw fa-lg" aria-hidden="true"></i> {{ $dashboard->dashboard_name }}</a></li>
+                                @endforeach
+                                <li role="presentation" class="divider"></li>
+                                <li>
+                                    <a href="{{ url('toggle_dashboard_editor') }}">
+                                    <i class="fa fa-bar-chart fa-fw fa-lg" aria-hidden="true"></i>
+                                    @if ($hide_dashboard_editor) @lang('Show Dashboard Editor')
+                                    @else @lang('Hide Dashboard Editor')
+                                    @endif</a>
+                                    <form id="toggle_dashboard_editor-form" action="{{ url('toggle_dashboard_editor') }}" method="POST" style="display: none;">
```
Comment (Jellyfrog) [inline resources/views/layouts/menu.blade.php:41]: What does this empty form do?

Comment (SourceDoctor) [reply resources/views/layouts/menu.blade.php:41]: it calls a controller method, similar to:
```
<form id="logout-form" action="{{ route('logout') }}" method="POST" style="display: none;">
{{ csrf_field() }}
</form>
```

File: app/Http/Controllers/LegacyController.php (hunk 2 above)
Comment (SourceDoctor) [reply app/Http/Controllers/LegacyController.php:55]: should 'front_page' be removed?

**Comment by Jellyfrog** [comment]
An idea, can't the menu item simply show/hide the editor with Javascript? Then we don't need any userperfs.

File: app/Http/Controllers/LegacyController.php (hunk 2 above)
Comment (SourceDoctor) [reply app/Http/Controllers/LegacyController.php]: removed it :-)

File: resources/views/layouts/menu.blade.php (hunk 3 above)
Comment (SourceDoctor) [reply resources/views/layouts/menu.blade.php]: ok, a little bit confused. Before i didn't get it to work without it, now i tried it without it and worked. ... removed now :-)

**Comment by SourceDoctor** [comment]
@Jellyfrog 
Target is to have it's configuration persistent and userspecific.

Also per default it should be seen. I think about new LibreNMS User :-)

**Comment by Jellyfrog** [comment]
Yes. But you only need check that in overview.blade.php, in the menu you simply have a Javascript `.toggle()`. That way we don't need to Add "hide_dashboard_editor" here and there

**Comment by SourceDoctor** [comment]
variable has to be in MenuComposer also, otherwise Blade based Views crashing because of missing this variable.
As i remember Menubar is parent of all sites ...
in Menu it swaps between 'show' and 'hide', so it's needed

**Comment by SourceDoctor** [comment]
@murrant @Jellyfrog 
something else to handle?

File: resources/views/layouts/menu.blade.php [hunk 4]
```diff
@@ -25,8 +25,23 @@
                                                  aria-hidden="true"></i> <span
                             class="hidden-sm">@lang('Overview')</span></a>
                     <ul class="dropdown-menu multi-level" role="menu">
-                        <li><a href="{{ url('overview') }}"><i class="fa fa-tv fa-fw fa-lg"
-                                                               aria-hidden="true"></i> @lang('Dashboard')</a></li>
+                        <li class="dropdown-submenu">
+                            <a><i class="fa fa-tv fa-fw fa-lg" aria-hidden="true"></i> @lang('Dashboard')</a>
+                            <ul class="dropdown-menu">
+                                @foreach($dashboards as $dashboard)
+                                <li><a href="overview?dashboard={{ $dashboard->dashboard_id }}"><i class="fa fa-tv fa-fw fa-lg" aria-hidden="true"></i> {{ $dashboard->dashboard_name }}</a></li>
```
Comment (Jellyfrog) [inline resources/views/layouts/menu.blade.php:32]: Should use route() instead of hard coded link

File: resources/views/layouts/menu.blade.php [hunk 5]
```diff
@@ -25,8 +25,23 @@
                                                  aria-hidden="true"></i> <span
                             class="hidden-sm">@lang('Overview')</span></a>
                     <ul class="dropdown-menu multi-level" role="menu">
-                        <li><a href="{{ url('overview') }}"><i class="fa fa-tv fa-fw fa-lg"
-                                                               aria-hidden="true"></i> @lang('Dashboard')</a></li>
+                        <li class="dropdown-submenu">
+                            <a><i class="fa fa-tv fa-fw fa-lg" aria-hidden="true"></i> @lang('Dashboard')</a>
```
Comment (Jellyfrog) [inline resources/views/layouts/menu.blade.php:29]: Missing href.

Comment (SourceDoctor) [reply resources/views/layouts/menu.blade.php:29]: was removed.
look on comment here:
[](https://github.com/librenms/librenms/pull/11321#issuecomment-603255057)

File: resources/views/layouts/menu.blade.php (hunk 4 above)
Comment (SourceDoctor) [reply resources/views/layouts/menu.blade.php]: changed

**Comment by SourceDoctor** [comment]
@Jellyfrog 
something else to change?

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-63-release-changelog-april-2020/11828/1

**Pull request merged**